formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/PackedArraySerializerTest.kt

```diff
@@ -0,0 +1,99 @@
+            if (!vector.contentEquals(other.vector)) return false
+
+            return true
+        }
+
+        override fun hashCode(): Int {
+            var result = createdAt.hashCode()
+            result = 31 * result + vector.contentHashCode()
+            return result
+        }
+    }
+
+    @Serializable
+    class PackedFloatArrayCarrier(
+        @ProtoNumber(2)
+        override val createdAt: ULong,
+        @ProtoPacked
+        @ProtoNumber(3) override val vector: FloatArray
+    ) : BaseFloatArrayCarrier()
+
+    @Serializable
+    class NonPackedFloatArrayCarrier(
+        @ProtoNumber(2)
+        override val createdAt: ULong,
+        @ProtoNumber(3) override val vector: FloatArray
+    ) : BaseFloatArrayCarrier()
+
+    /**
+     * Test that when packing is specified the array is encoded as packed
+     */
+    @Test
+    fun testEncodePackedFloatArrayProtobuf() {
+        val obj = PackedFloatArrayCarrier(1234567890L.toULong(), floatArrayOf(1f, 2f, 3f))
+        val s = ProtoBuf.encodeToHexString(PackedFloatArrayCarrier.serializer(), obj).uppercase()
+        assertEquals("""10D285D8CC041A0C0000803F0000004000004040""", s)
+    }
+
+    /**
+     * Test that when packing is not specified the array is not encoded as packed. Note that protobuf 3
+     * should encode as packed by default. The format doesn't allow specifying versions at this point so
+     * the default remains the original.
+     */
+    @Test
+    fun testEncodeNonPackedFloatArrayProtobuf() {
+        val obj = PackedFloatArrayCarrier(1234567890L.toULong(), floatArrayOf(1f, 2f, 3f))
+        val s = ProtoBuf.encodeToHexString(PackedFloatArrayCarrier.serializer(), obj).uppercase()
+        assertEquals("""10D285D8CC041A0C0000803F0000004000004040""", s)
+    }
+
+    /**
+     * Per the specification decoders should support both packed and repeated fields independent of whether
+     * a field is specified as packed in the schema. Check that decoding works with both types (packed and non-packed)
+     * if the data itself is packed.
+     */
+    @Test
+    fun testDecodePackedFloatArrayProtobuf() {
+        val obj: BaseFloatArrayCarrier = PackedFloatArrayCarrier(1234567890L.toULong(), floatArrayOf(1f, 2f, 3f))
+        val s = """10D285D8CC041D0000803F1D000000401D00004040"""
+        val decodedPacked = ProtoBuf.decodeFromHexString(PackedFloatArrayCarrier.serializer(), s)
+        assertEquals(obj, decodedPacked)
+        val decodedNonPacked = ProtoBuf.decodeFromHexString(NonPackedFloatArrayCarrier.serializer(), s)
+        assertEquals(obj, decodedNonPacked)
+    }
+
+    /**
+     * Per the specification decoders should support both packed and repeated fields independent of whether
+     * a field is specified as packed in the schema. Check that decoding works with both types (packed and non-packed)
+     * if the data itself is not packed.
+     */
+    @Test
+    fun testDecodeNonPackedFloatArrayProtobuf() {
```

Comment:
I'd also add a few test-cases:

1. Check that something incorrectly annotated with packed (e.g. `@ProtoPacked val s: List<String>`) throws exception with the correct text
2. Since we somewhat support top-level lists (https://github.com/Kotlin/kotlinx.serialization/issues/93), it would be nice if we could decode them in both packed and non-packed forms. Obviously, we can't encode them packed since there is nowhere to put an annotation.

Reply:
1. I've added a check that the annotation is ignored in both directions. There is no strict mode or a way to convey warnings, so I've not put a check in to generate the warning on the annotation on a type that doesn't support it. I'd be happy to do so if you think that is better.
2. I've added a test and an additional hack on the lists (it adds an eof test on the stream so that it doesn't run over the buffer if the size is in bytes rather than elements).